docs/data-management/edf-to-bids.md

```diff
@@ -0,0 +1,634 @@
+
+??? tip "*PyEDFRead* must be installed"
+
+    Make sure *PyEDFRead* is installed on the computer where conversion will be executed, as described in [the software annex](data-collection/notes-software.md#conversion-of-et-recordings-into-bids).
+
+- [ ] Open a run's EDF file:
+
+    ``` Python
+    from pyedfread import edf
+
+    recording, events, messages = edf.pread(str(DATA_PATH / "example-run.edf"), trial_marker=b"")
+    ```
+
+    As a result, thre *Pandas*' dataframes will be obtained:
+
+    * `recoding` contains the actual ET data,
+    * `events` contains information of events reported by the device (most substantially these are fixations, saccades, and blinks), and
+    * `messages` with the logged textual messages sent to the device.
+
+## Processing the *Pandas* dataframes generated by *PyEDFRead*
+
+The code snippets are derived from the [`EyeTrackingRun` object](../code/eyetracking/eyetrackingrun.py).
+
+!!! warning "Disclamer: despite this code attempting to be general, some parts focus on single-eye recordings"
+
+### Parsing the `messages` dataframe
+
+First, we clean the `messages` dataframe, as it may have headers ending with space and drop duplicate rows:
+
+``` Python
+messages = messages.rename(
+    columns={c: c.strip() for c in messages.columns.values}
+).drop_duplicates()
+```
+
+**Extract and reserve calibration messages**.
+EyeLink reports the calibration information as messages, which are marked up with a `!CAL` prefix.
+We extract them into a separate variable to process them at a later time:
+
+```Python
+# Extract calibration headers
+_cal_hdr = messages.trialid.str.startswith("!CAL")
+calibration = messages[_cal_hdr]
+messages = messages.drop(messages.index[_cal_hdr])
+```
+
+**Extracting the `StartTime` and `StopTime` metadata**.
+We then search for messages signaling the start and stop run events.
+These events are stored in the JSON sidecar file annotating our `_eyetrack.tsv.gz` file.
+In our *Psychopy* experiment files, we configure messages be sent to the ET at these events.
+The message changes a little with the RSfMRI experiment, but are usually like `hello <taskname>` and `bye <taskname>`, which are stored in the variables `message_first_trigger` and `message_last_trigger` variables, respectively.
+In the last line, we clean up these used-up lines from the dataframe.
+
+``` Python
+# Find Start time
+start_rows = messages.trialid.str.contains(
+    message_first_trigger, case=False, regex=True
+)
+stop_rows = messages.trialid.str.contains(
+    message_last_trigger, case=False, regex=True
+)
+
+# Extract calibration headers
+_cal_hdr = messages.trialid.str.startswith("!CAL")
+calibration = messages[_cal_hdr]
+messages = messages.drop(messages.index[_cal_hdr])
+
+# Pick the LAST of the start messages
+metadata["StartTime"] = (
+    int(messages[start_rows].trialid_time.values[-1])
```

Comment:
Why taking the last one?

Reply:
Because it is possible that the run was started, the message sent and then cancelled and re-launched a second run. In that case, it is the last message what you want to follow. Does this make sense?

NOTE - we need to check that in that scenario of cancelling the run we do not get the `bye` message.

Reply:
It does make sense, but when we restart the task in PsychoPy, it creates a new EDF file. Therefore, the scenario where two runs are in the same file should not occur. However, in the case of the resting state, we have multiple messages that begin with 'start,' so with this change we will have to be careful and specify the entire string 'start movie,' not just 'start.'

Reply:
Indeed.